### PR TITLE
Use ITK_TEMPLATE_EXPORT for templated classes in elastix

### PR DESCRIPTION
**Pull request by ViktorvdValk**

### CODE REVIEW

**Comment by N-Dekker** [comment]
Hi Viktor, can you please make it one commit? (By squashing locally + `git push --force-with-lease`).

**Comment by ViktorvdValk** [comment]
Yes, if it passes, I will.

**Comment by N-Dekker** [comment]
I guess it's a follow-up to pull request https://github.com/SuperElastix/elastix/pull/412 ('fixed_template_export_macro_in_classes') by @thewtex right?

File: Components/Transforms/TranslationStackTransform/itkAdvancedTranslationTransform.h [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+/*=========================================================================
+
+  Program:   Insight Segmentation & Registration Toolkit
+  Module:    $RCSfile: itkAdvancedTranslationTransform.h,v $
```
Comment (N-Dekker) [inline Components/Transforms/TranslationStackTransform/itkAdvancedTranslationTransform.h:21]: This file "Components/Transforms/TranslationStackTransform/itkAdvancedTranslationTransform.h" was removed with pull request https://github.com/SuperElastix/elastix/pull/447 commit 7fdca9705a29541590e22bde7d8234536e5f5d20  Can you please keep it removed?

File: .github/workflows/ElastixGitHubActions.yml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 name: Elastix
 
 on: [push, pull_request]
-  
+
```
Comment (N-Dekker) [inline .github/workflows/ElastixGitHubActions.yml:4]: Preferably don't modify  .github/workflows/ElastixGitHubActions.yml with this pull request, as it seems unrelated.

**Comment by ViktorvdValk** [comment]
@N-Dekker Do you want to check this PR or shall I merge it?

File: Common/Transforms/itkEulerTransform.h [hunk 3]
```diff
@@ -46,7 +46,7 @@ class EulerGroup
  */
 
 template <>
-class EulerGroup<2>
+class ITK_TEMPLATE_EXPORT EulerGroup<2>
```
Comment (N-Dekker) [inline Common/Transforms/itkEulerTransform.h:49]: I wonder if `ITK_TEMPLATE_EXPORT` should also be applied to explicit (full) template specializations, like `EulerGroup<2>`...

File: Components/Transforms/TranslationStackTransform/itkAdvancedTranslationTransform.h (hunk 1 above)
Comment (N-Dekker) [reply Components/Transforms/TranslationStackTransform/itkAdvancedTranslationTransform.h]: Thanks, I see you addressed my comment on itkAdvancedTranslationTransform.h 👍

File: .github/workflows/ElastixGitHubActions.yml (hunk 2 above)
Comment (N-Dekker) [reply .github/workflows/ElastixGitHubActions.yml]: Thanks, I see you addressed my comment on ElastixGitHubActions.yml 👍

File: Common/TypeList.h [hunk 4]
```diff
@@ -85,7 +85,7 @@ namespace typelist
  *     Tail (second element, can be another typelist)
  */
 template <typename H, typename T>
-struct TypeList
+struct ITK_TEMPLATE_EXPORT TypeList
```
Comment (N-Dekker) [inline Common/TypeList.h:88]: I wonder if `ITK_TEMPLATE_EXPORT` is really useful for the templates in Common/TypeList.h, as those templates do not have any member functions.

**Comment by N-Dekker** [comment]
Could you possibly still extend the commit message to clarify the purpose of the commit? Also mention that it's a follow-up to

Revision: df2a3ce40c34e94abf0613e5f7f113bf1ca97b0f
Author: ViktorvdValk
Date: 26 Nov 2020 14:43:16
Message: COMP: Fix template export macro in classes

**Comment by ViktorvdValk** [comment]
@N-Dekker Shall I leave ITK_TEMPLATE_EXPORT on the templates you just mentioned? Since there is currently no way to verify if the macros are needed or not.

**Comment by N-Dekker** [comment]
For the record,

Related ITK commits by MayeulChassagnard and Matt McCormick, 5 Dec 2016:

 - https://github.com/InsightSoftwareConsortium/ITK/commit/be14aef1ec8646982fa717b7a88b8f1b60558717 "BUG: Add ITK_TEMPLATE_EXPORT to template classes"

 - https://github.com/InsightSoftwareConsortium/ITK/commit/8bc35937c09657002bfd05517a0edf5dec736695 "ENH: Add tests to demonstrate OSX dynamic_cast failure"

**Comment by ViktorvdValk** [comment]
I'll extend the merge message and clarify the purpose of the commit, once I merge the PR, is that oke?

File: Common/TypeList.h (hunk 4 above)
Comment (N-Dekker) [reply Common/TypeList.h:88]: I think the ITK_TEMPLATE_EXPORT is only needed for structs and classes that have `virtual` member functions (the types that may be used for `dynamic_cast`). So they should not be necessary for the templates in Common/TypeList.h

**Comment by N-Dekker** [comment]
> I'll extend the merge message and clarify the purpose of the commit, once I merge the PR, is that oke?

OK, no problem, but why not just do a `git amend` + `git push --force-with-lease` from your pull-request-branch?

**Review by N-Dekker — APPROVED** [review]
Thanks, Viktor! 👍
